### PR TITLE
GitHub actions

### PR DESCRIPTION
**Pull request by br648**
Replace Travis CI with GitHub Actions.

The following third party actions has to be reviewed and approved by two dev before the this PR can be completed:

https://github.com/toko-bifrost/ms-teams-deploy-card/releases/tag/3.1.2

Suggested reviewers:

Rob Beer - Reviewed
Evan Siroky - Reviewed (18/02/2021)

Note the following issues:

1) I'm seeing double builds under GitHub actions. This is probably related to push and pull trigger on build.

### CODE REVIEW

File: .github/workflows/maven-ci.yml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#      - master
+#      - dev-1.x
+#      - dev-2.x
+#      - ibi-dev
+#  pull_request:
+#    branches:
+#      - master
+#      - dev-1.x
+#      - dev-2.x
+#      - ibi-dev
```
Comment (landonreed) [inline .github/workflows/maven-ci.yml]: I think we can remove this commented out branch filtering.

File: .github/workflows/maven-ci.yml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  build:
+    runs-on: ubuntu-latest
+
+    env:
+      JAVA_OPTS: -Xmx2g
+      # TODO Remove if not required.
+      OSSRH_JIRA_USERNAME: ${{ secrets.OSSRH_JIRA_USERNAME }}
+      OSSRH_JIRA_PASSWORD: ${{ secrets.OSSRH_JIRA_PASSWORD }}
+      GPG_KEY_NAME: ${{ secrets.GPG_KEY_NAME }}
+      GPG_PASSPHRASE: ${{ secrets.GPG_PASSPHRASE }}
```
Comment (landonreed) [inline .github/workflows/maven-ci.yml]: Yea, I think it's unneeded (only for the main repo).

File: .github/workflows/maven-ci.yml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        if: github.event_name == 'push' && env.GITHUB_HEAD_REF_SLUG == 'ibi-dev'
+        uses: jakejarvis/s3-sync-action@master
+        with:
+          args: --recursive --acl public-read
+        env:
+          AWS_S3_BUCKET: otp-docs
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          SOURCE_DIR: 'target/mkdocs'
+      # TODO: Push notification ot Teams.
```
Comment (landonreed) [inline .github/workflows/maven-ci.yml]: typo: ot

File: .github/workflows/maven-ci.yml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+          path: ~/.m2
+          # TODO Fix error message 'Cache not found for input keys: maven-local-repo'
+          key: maven-local-repo
+      # So that the branch name can be reference via $GITHUB_HEAD_REF_SLUG.
+      - name: Inject slug/short variables
+        uses: rlespinasse/github-slug-action@v3.x
+      # Define AWS profile credentials for use by subsequent steps.
+      - name: Add profile credentials to ~/.aws/credentials
+        if: github.event_name == 'push'
+        run: ./scripts/add-aws-credentials.sh
```
Comment (landonreed) [inline .github/workflows/maven-ci.yml:37]: I don't think this step is needed since we're using the `jakejarvis/s3-sync-action` action.

File: scripts/prep-build-for-deploy-from-github-actions.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+# `git describe --tags --always HEAD` returns a string of the form v0.0.0-52-ge10d02d.
+# It assumes you have pushed a tag on a commit on github (e.g. a commit on the dev branch).
+# If for some reason git can't find a tag, fallback with --always to a commit sha.
+export JAR_VERSION=$(git describe --tags --always HEAD)
+echo $JAR_VERSION
+# Create a deployment folder, and a folder for the branch.
+mkdir deploy
+# Add the JAR file.
+cp target/*-shaded.jar deploy/otp-$JAR_VERSION.jar
+cp target/*-shaded.jar deploy/otp-latest-$BRANCH.jar
```
Comment (landonreed) [inline scripts/prep-build-for-deploy-from-github-actions.sh:19]: Issue w/ BRANCH variable: https://github.com/ibi-group/OpenTripPlanner/pull/50/checks?check_run_id=1792953482#step:10:31

File: .github/workflows/maven-ci.yml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+      # Build and test deployment.
+      - name: Build and Test
+#        TODO: Remove skip tests. Temporary change to allow testing of later steps.
+        run: mvn clean verify --batch-mode --show-version --update-snapshots -Dgpg.skip -DskipTests
+#        TODO: Confirm if OSSRH/GPG secrets are required.
+#        run: mvn clean verify --settings maven-settings.xml --batch-mode --show-version --update-snapshots -Dgpg.skip
```
Comment (landonreed) [inline .github/workflows/maven-ci.yml]: Nope, I think we can remove.

**Review by landonreed — CHANGES_REQUESTED** [review]
Just a few changes.

File: .github/workflows/maven-ci.yml (hunk 1 above)
Comment (br648) [reply .github/workflows/maven-ci.yml]: @landonreed I'm seeing double builds under GitHub actions as noted in the original comment, perhaps there is something in this?!

File: .github/workflows/maven-ci.yml (hunk 2 above)
Comment (br648) [reply .github/workflows/maven-ci.yml]: Removed.

File: .github/workflows/maven-ci.yml (hunk 3 above)
Comment (br648) [reply .github/workflows/maven-ci.yml]: Updated.

File: .github/workflows/maven-ci.yml (hunk 4 above)
Comment (br648) [reply .github/workflows/maven-ci.yml:37]: Step removed.

Comment (br648) [reply .github/workflows/maven-ci.yml:37]: Step removed.

File: scripts/prep-build-for-deploy-from-github-actions.sh (hunk 5 above)
Comment (br648) [reply scripts/prep-build-for-deploy-from-github-actions.sh:19]: Fixed

File: .github/workflows/maven-ci.yml (hunk 6 above)
Comment (br648) [reply .github/workflows/maven-ci.yml]: Removed.

File: .github/workflows/maven-ci.yml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      - name: Copy build files to the otp-repo bucket
+        if: github.event_name == 'push'
+        uses: jakejarvis/s3-sync-action@master
+        with:
+          args: --acl public-read
+        env:
+          AWS_S3_BUCKET: otp-repo
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          SOURCE_DIR: 'deploy'
```
Comment (evansiroky) [inline .github/workflows/maven-ci.yml]: I realize that this is nice and easy with just a bit of config, but this is a security risk to send AWS credentials to a third party GitHub action. I would only do this with something built specifically from either GitHub or Amazon. The underlying script that this plugin uses is quite simple, so I don't think it's worth the risk of including this. It'd probably be best to simply create a script in this repo that handles this so there is no risk of credentials being misused by a third party.

File: .github/workflows/maven-ci.yml [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      - name: Upload docs to IBI s3 buckets & cloudfront
+        if: github.event_name == 'push' && env.GITHUB_HEAD_REF_SLUG == 'ibi-dev'
+        uses: jakejarvis/s3-sync-action@master
+        with:
+          args: --acl public-read
+        env:
+          AWS_S3_BUCKET: otp-docs.ibi-transit.com
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          SOURCE_DIR: 'target/mkdocs'
```
Comment (evansiroky) [inline .github/workflows/maven-ci.yml]: Same comment as https://github.com/ibi-group/OpenTripPlanner/pull/50/files#r568920120

File: .github/workflows/maven-ci.yml [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      # Push notification to Teams.
+      - name: Notify us of the build status on MS Teams
+        uses: toko-bifrost/ms-teams-deploy-card@master #  or "./" if in a local set-up
+        if: always()
+        with:
+          github-token: ${{ github.token }}
+          # TODO Add webhook to GitHub secrets
+          # webhook_url: ${{ secrets.MSTEAMS_WEBHOOK }}
+          webhook_url: https://outlook.office.com/webhook/03fa4a79-572f-4c68-b756-e4e851d0215a@9093f1a3-8771-4fb7-8596-d51eeef18cda/TravisCI/0aec73ecb8eb4ad696ea
```
Comment (evansiroky) [inline .github/workflows/maven-ci.yml]: Another security risk here with allowing the GitHub token to be used by a third party action. If there is another action that is made by either GitHub or Microsoft, then that is fine, but getting an MS teams notification doesn't seem worth the security risk of providing a third party a GitHub token. I'd recommend this be removed.

**Review by evansiroky — CHANGES_REQUESTED** [review]
Please refactor to minimize security risks as noted.

File: .github/workflows/maven-ci.yml (hunk 4 above)
Comment (evansiroky) [reply .github/workflows/maven-ci.yml:37]: Unresolving. Giving AWS credentials to a 3rd party is an unnecessary security risk.

File: .github/workflows/maven-ci.yml (hunk 7 above)
Comment (br648) [reply .github/workflows/maven-ci.yml]: Reverted back to using a script.

File: .github/workflows/maven-ci.yml (hunk 8 above)
Comment (br648) [reply .github/workflows/maven-ci.yml]: Revert back to using a script

File: .github/workflows/maven-ci.yml (hunk 9 above)
Comment (br648) [reply .github/workflows/maven-ci.yml]: @evansiroky Plenty to choose from in the GitHub Marketplace, but none appear to be made by either GitHub or Microsoft. I think a call can be made using curl, but I'm not sure if I should invest the time in doing that?

File: .github/workflows/maven-ci.yml [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    env:
+      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+      AWS_PROFILE: default
+      AWS_REGION: us-east-1
+      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+      MAVEN_OPTS: -Xmx2g
```
Comment (evansiroky) [inline .github/workflows/maven-ci.yml]: This makes available our AWS access keys to all steps and thus third party applications. If possible, can these be included only in the individual steps that need this logic?

File: .github/workflows/maven-ci.yml (hunk 4 above)
Comment (br648) [reply .github/workflows/maven-ci.yml:37]: Now using script and AWS creds are only available to this step.

**Comment by br648** [comment]
@evansiroky When you get the chance, can you please review the third party action linked in the description? Once you are happy, I will then uncomment and test the action.

File: .github/workflows/maven-ci.yml (hunk 10 above)
Comment (br648) [reply .github/workflows/maven-ci.yml]: AWS creds moved into individual step.

File: .github/workflows/maven-ci.yml (hunk 9 above)
Comment (br648) [reply .github/workflows/maven-ci.yml]: The plan is to review third party actions and use if suitable.

**Comment by evansiroky** [comment]
> @evansiroky When you get the chance, can you please review the third party action linked in the description? Once you are happy, I will then uncomment and test the action.

toko-bifrost/ms-teams-deploy-card@3.1.2 seems fine to me. It looks like the Github token is only used to get info about the commit and workflows, so seems ok.

**Comment by br648** [comment]
@landonreed this is now ready for a second review. I'll get @evansiroky to review once he is available again.

File: src/main/java/org/opentripplanner/graph_builder/module/ned/NEDDownloader.java [hunk 11]
```diff
@@ -545,7 +545,7 @@ public String getPrefix(String namespace) {
         }
     }
 
-    public Iterator<?> getPrefixes(String namespace) {
+    public Iterator getPrefixes(String namespace) {
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/graph_builder/module/ned/NEDDownloader.java]: Is this change actually needed to make things build/tests pass?

**Review by landonreed — APPROVED** [review]
LGTM. Just one non-blocking question about a potentially unnecessary file change in `src/`.

File: src/main/java/org/opentripplanner/graph_builder/module/ned/NEDDownloader.java (hunk 11 above)
Comment (br648) [reply src/main/java/org/opentripplanner/graph_builder/module/ned/NEDDownloader.java]: @landonreed - JDK version related. 

This fails:
Project SDK: OpenJDK 11
Module (otp): Oracle JDK 1.8_191

This works:
Project SDK: Oracle JDK 1.8_191
Module (otp): Oracle JDK 1.8_191

I'm not sure why it is not using the module specific JDK. Anyway, assuming this project is still 1.8, I have removed this alteration.

**Review by evansiroky — APPROVED** [review]
